// libvita/test2/test3.js
var tester = require('../tester')
	,addon = tester.addon;

var test3Num = 93
	,voltMargin = .11
	,dacMargin = .03;

var test3 = {
	name: 'AN2012->AN2024 Timing Test',
	description: 'This test sets a voltage on one of the DACs on 2012 and reads it on one of the ADC channels of 2024.  Runs several tests from .7V to 10V.  Measures the timing of output change and agent callbacks.',
	sync: 'test2', // run after test2 instead of asynchronously
	init: function(callback) {
		addon.open('usb:CLASS_AN2012/66666666', function(err, result) {
			if (err) console.log('Error:an2012 failed to open, ' + err);
			else an2012 = result;
			addon.open('usb:CLASS_AN2024/00108158', function(err, result) {
				if (err) console.log('Error:an2024 failed to open, ' + err);
				else an2024 = result;
				callback(null);
			});
		});
	},
	close: function(callback) {
		if (an2012) an2012.close();
		if (an2024) an2024.close();
		return callback(null);
	},
	run: function(callback) {
		var outputArr = []
			,index = 0
			,changeResults = []
			,agentResults = []
			,agentCount = 0
			,currentOutput
			,startTimeChange
			,startTimeCallback
			,self = this
			,oldV = -1
			,testRuns = test3Num;

		console.log('running test3');
		for (var i = 0; i < testRuns; i++) {
			outputArr.push(i * 100 + 700);
		}
		currentOutput = outputArr[index++];

		an2012.setOutputs(1, 1, currentOutput, function(err) {
			if (err) return callback("Error: setting outputs on 2024, " + err);

			startTimeChange = process.hrtime();
			startTimeCallback = process.hrtime();
			self.append('set voltage ' + currentOutput);

			an2024.startAgent(2, 1, 10, function(err, data, agentId) {
				if (err) console.log("Error: agent on 2024, " + err);

				else if (index == testRuns + 1) {// stop condition
					an2024.stopAgent(agentId);
					var average = self.averageArr(changeResults);
					self.append('****************Stopped Agent 3*******************');
					self.append("agent3 change data: " + changeResults);
					self.append("agent3 callback data: " + agentResults);
					self.append("\tagent3 count: " + agentCount);
					self.append("\tagent3 changes: " + testRuns);
					self.appendResults("Agent3 average change time: " + average + " ms");
					self.appendResults("Agent3 average callback time: " + self.averageArr(agentResults) + " ms");
					console.log('stopped test3');
					return callback(null, self);
				} else { //iterate
					var agentTime = process.hrtime(startTimeCallback);
					startTimeCallback = process.hrtime();
					agentResults.push(self.toMs(agentTime));
					data *= 10;

					self.append("\tAgent 3 data: " + data);
					agentCount++;

					if (self.inRange(data, currentOutput, voltMargin) && data != oldV) {//output changed
						var finishTime = self.toMs(process.hrtime(startTimeChange));
						changeResults.push(finishTime);
						self.append('got voltage: ' + data);

						oldV = data;
						currentOutput = outputArr[index++];
						an2012.setOutputs(1, 1, currentOutput, function(err) {
							if (err) return callback("Error: setting outputs on 2024, " + err);
							startTimeChange = process.hrtime();
							if (index < testRuns + 1) self.append('set voltage ' + currentOutput);
						});
					}
				}
			});
		});
	}
};

exports.test3 = test3;
